### ISSUE TITLE
Test last stable vpp releases and do update if no problems occurred

### ISSUE THREAD

**Issue by denis-tingaikin** [networkservicemesh/govpp#24]
Plan:

1. Check and test all stable releases since our current vpp version. For testing, use forwarder-vpp docker tests and NSM integration tests.
2. If no problem occurred, just do an update
3. Otherwise, provide a table with the next information: vpp version, which tests don't work
4. [Optional] Find a root cause of problems and reproduce it via pure VPP.
5. [Optional] Report found problems to vpp
6. [Optional] Commit fixes.

**Comment by NikitaSkrynnik** [comment]
NSM doesn't work with VPP v24.02 and Calico stilll uses v24.02-rc0. I think it's reasonable to wait for a new release from Calico and then update VPP in NSM

**Comment by Ex4amp1e** [comment]
Plan:
- [ ] vpp 24.06
- [ ] vpp 24.02

Decomposition:

- [x] Run govpp docker tests without changes - 30m
- [x] Build govpp based on vpp 24.06 - 4h
- [x] Run cmd-forwarder-vpp tests based on vpp 24.06 - 2h
- [ ] Build govpp based on vpp 24.02 - 4h
- [ ] Run cmd-forwarder-vpp tests based on vpp 24.06 - 2h

**Comment by Ex4amp1e** [comment]
Current results:
cmd-forwarder-vpp docker tests fail on vpp v24.06:
```
            --- FAIL: TestForwarderTestSuite/TestCombinations/Ethernet/Memif (2.53s)
                --- FAIL: TestForwarderTestSuite/TestCombinations/Ethernet/Memif/Kernel (0.66s)
                --- FAIL: TestForwarderTestSuite/TestCombinations/Ethernet/Memif/Memif (0.92s)
                --- FAIL: TestForwarderTestSuite/TestCombinations/Ethernet/Memif/Vxlan (0.95s)
```
Note: ipv6 pings fail
Full docker log: [24.06.new.txt](https://github.com/user-attachments/files/17379165/24.06.new.txt)

vpp v24.02 has been tested and NSM doesn't work https://github.com/networkservicemesh/govpp/issues/24#issuecomment-1998889081

**Possible options:**
Option 1: Find  the root cause why Ethernet/Memif tests do not work in vpp code
Option 2: Wait vpp v24.10 (looks like it should be released at the end of Oct 2024) and test it

**Comment by denis-tingaikin** [comment]
cc @szvincze

**Comment by elajkat** [comment]
24.10 is out, see the re-no page: https://s3-docs.fd.io/vpp/24.10/aboutvpp/releasenotes/v24.10.html
(https://gerrit.fd.io/r/c/vpp/+/41646 )

**Comment by elajkat** [comment]
I am not sure I use the govpp api diff in correct way but the API diff for v23.06 to v24.06:
https://pastebin.com/jQjPWdWN

(I used the command: 
$ cd ~/temp; git clone https://github.com/FDio/vpp; git checkout v24.06; cd ~/vpp; git checkout v23.06; govpp vppapi diff --against ~/temp/vpp/ 
)
Perhaps worth to check first the API diff why those tests are failing.

(The API diff for 23.06=>24.10 is here: https://pastebin.com/L52EqTzX )

**Comment by NikitaSkrynnik** [comment]
Current plan:
- [ ] Update `vpp` to the latest release
- [ ] Run docker tests in `cmd-forwarder-vpp`
- [ ] If docker tests work then run all tests
- [ ] If docker tests don't work then reproduce the same setup from the failed tests on plain `vpp`

**Comment by NikitaSkrynnik** [comment]
We've tested `vpp v24.10` on our CI and docker tests. There is a problem with `memif` interfaces. Particularly tests with `vpp` endpoint don't work.

Current plan:
- [ ] test `kernel2memif` scenario on plain `vpp`
- [ ] if it works then find the difference between our setup and the plain `vpp` setup

**Comment by elajkat** [comment]
Hi, you reproduced the failure (memif with VPP 24.10) with the integration-k8s-kind environment? I try to reproduce it in my local env.

**Comment by NikitaSkrynnik** [comment]
Yes, I used our integration tests. Tests work if I launch them separately with different forwarders. But if I run the whole `basic` suite all of them fail except the first one. It looks like we don't clean resources on forwarder properly. The same behaviour is in docker tests. Plain `vpp` memif tests work correctly. Current plan:
- [ ] debug docker tests

**Comment by NikitaSkrynnik** [comment]
Current state:
1. Docker tests has been fixed after disabling `direct memif` feature in `cmd-forwarder-vpp`: [code](https://github.com/networkservicemesh/sdk-vpp/blob/main/pkg/networkservice/chains/forwarder/server.go#L132)
2. Integration-tests don't work even without `direct memif`. Debugging of `Kernel2Memif` has shown that the problem is with `memif` interfaces. `cmd-nse-vpp` shows error `no free tx slots` after several successful pings. Manual setup of `Kernel2Memif` with two plain `vpp` instances doesn't have this problem.

**Comment by elajkat** [comment]
Most probably this was the 1st patch in VPP that broke how NSM uses it:
https://gerrit.fd.io/r/c/vpp/+/39621

First inspection shows that tuning the buffer section values in VPP can solve the failing tests issue, see:
https://fd.io/docs/vpp/v2101/gettingstarted/users/configuring/startup.html#the-buffers-section

In my local env I was able to execute all the single_tests both with vpp2402 and vpp2410 with these values:
buffers{
   buffers-per-numa 64000
   default data-size 2048
}

More digging is necessary to see why a numa buffer size affects this behaviour even when we do not use numa in our test env, and a mail to VPP folks to see the background and possible advises from them.

**Comment by szvincze** [comment]
It seems the allocation of the buffer pools have been improved in `v24.x` as @elajkat referred the first commit in this area.
In `VPP v23.10` the number of buffers was simply calculated as `number of pages * number of allocations per page` but the buffers that would span across page boundary were not skipped. This way `vppctl show buffers` printed the configured number of pools even if not all of them were available. `VPP v24.10` allocates the buffer pools with the mentioned consideration.

You can see below that the same configuration results different number of buffers on the two releases.

VPP v23.10:
```
# vppctl sh buffers
Pool Name            Index NUMA  Size  Data Size  Total  Avail  Cached   Used  
default-numa-0         0     0   4032     3776    32768  32255    181     332  
```
VPP v24.10:
```
# vppctl sh buffers
Pool Name            Index NUMA  Size  Data Size  Total  Avail  Cached   Used  
default-numa-0         0     0   4032     3776     520     8      255     257  
```
Having just 8 available buffers on `VPP 24.10` caused the unstable behavior of the NSM integration tests. If we just run a single test that would be successful based on the actually available buffers but running the test parallel ends up failed executions.

Of course, one could just double the `buffers-per-numa`'s value but that will only double the `Total` (520). However going back to the default `data-size` (2048 bytes) will result much higher number of buffer pools because of the contiguous memory chunks, see below:
```
Pool Name            Index NUMA  Size  Data Size  Total  Avail  Cached   Used  
default-numa-0         0     0   2304     2048    23912  23400    190     322 
```
So, I would propose to change the `default data-size` configuration in [vpphelper](https://github.com/networkservicemesh/vpphelper/blob/main/constants.go#L24C2-L24C20) to `2048` instead of the current `3776` and run a thorough test with it.

**Comment by szvincze** [comment]
The uplift was released in NSM v1.14.3. There is a known issue with memif, for details please see the release notes document:
https://networkservicemesh.io/docs/releases/v1.14.3/#nsm-release-v1143
